---
# Data Science and Artificial Intelliegence Practicum
## 5-modul. Machine Learning
---

**Original Notebook -> https://jovian.ai/anvarnarz/05-ml-04-machinelearning**

## 5.4 - Machine Learning

### 1\. Data Preparation

#### Importing libraries for data analysis

In [1]:
# importing libraries
import numpy as np
import pandas as pd

#### Loading dataset

In [2]:
# loading dataset
URL = "https://github.com/ageron/handson-ml2/blob/master/datasets/housing/housing.csv?raw=true"
df = pd.read_csv(URL)

df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

#### Splitting data into `test_set` and `train_set`

In [3]:
df.shape

(20640, 10)

In [4]:
from sklearn.model_selection import train_test_split
train_set, test_set = train_test_split(df, test_size=0.2, random_state=42)

X_train = train_set.drop('median_house_value', axis=1)
y = train_set['median_house_value'].copy()

X_num = X_train.drop('ocean_proximity', axis=1)

### 2\. Building Pipeline

In [5]:
from sklearn.base import BaseEstimator, TransformerMixin

# indices of columns that we need
rooms_idx, bedrooms_idx, population_idx, households_idx = 3, 4, 5, 6

class CombinedAttributesAdder(BaseEstimator, TransformerMixin):
    
    def __init__(self, add_bedrooms_per_room=True):
        self.add_bedrooms_per_room = add_bedrooms_per_room
        
    def fit(self, X, y=None):
        return self  # our function is only a transformer (not an estimator)
    
    def transform(self, X):
        rooms_per_household = X[:, rooms_idx] / X[:, households_idx]
        population_per_household = X[:, population_idx] / X[:, households_idx]
        if self.add_bedrooms_per_room:  # add_bedrooms_per_room column is optional
            bedrooms_per_room = X[:, bedrooms_idx] / X[:, rooms_idx]
            return np.c_[X, rooms_per_household, population_per_household, bedrooms_per_room]
        else:
            return np.c_[X, rooms_per_household, population_per_household]

#### Pipeline for *numerical features*

In [6]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, StandardScaler

num_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('attribs_adder', CombinedAttributesAdder(add_bedrooms_per_room=True)),
    ('std_scaler', StandardScaler())
])

#### Pipeline for *categorical features*

In [7]:
from sklearn.compose import ColumnTransformer

num_attribs = list(X_num)
cat_attribs = ['ocean_proximity']

full_pipeline = ColumnTransformer([
    ('num', num_pipeline, num_attribs),
    ('cat', OneHotEncoder(), cat_attribs)
])

The final and complete pipeline(`full_pipeline`) is ready.

Using pipeline: call `.fit_transform()` method.

In [8]:
X_prepared = full_pipeline.fit_transform(X_train)
X_prepared

array([[ 1.27258656, -1.3728112 ,  0.34849025, ...,  0.        ,
         0.        ,  1.        ],
       [ 0.70916212, -0.87669601,  1.61811813, ...,  0.        ,
         0.        ,  1.        ],
       [-0.44760309, -0.46014647, -1.95271028, ...,  0.        ,
         0.        ,  1.        ],
       ...,
       [ 0.59946887, -0.75500738,  0.58654547, ...,  0.        ,
         0.        ,  0.        ],
       [-1.18553953,  0.90651045, -1.07984112, ...,  0.        ,
         0.        ,  0.        ],
       [-1.41489815,  0.99543676,  1.85617335, ...,  0.        ,
         1.        ,  0.        ]])

Dataset is read for Machine Learning.

### 3\. Machine Learning

Our goal is *prediction*, for this there are several Machine Learning algorithms.

#### Linear Regression

**`sklearn.linear_model.LinearRegression`** - Ordinary least squares Linear Regression.

In [9]:
from sklearn.linear_model import LinearRegression

LR_model = LinearRegression()

`LinearRegression` is an *estimator*. Estimator receives data and *learns* to predict by `fit` method.

In [10]:
LR_model.fit(X_prepared, y)

LinearRegression()

Linear regression model is ready!

How can we test the model? Let's feed a row from the `housing` dataset to the model and compare the result with the existing result (label).

In [11]:
# choosing 5 random sample rows
test_data = X_train.sample(5)
test_data

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
18330    -122.14     37.45                48.0       2074.0           297.0   
3139     -118.17     34.87                 9.0       1507.0           293.0   
11553    -117.98     33.74                29.0       3443.0           635.0   
272      -122.19     37.78                52.0       1070.0           193.0   
8079     -118.20     33.82                43.0       1758.0           347.0   

       population  households  median_income ocean_proximity  
18330       700.0       279.0         8.7051        NEAR BAY  
3139        761.0       278.0         3.0184          INLAND  
11553      2257.0       620.0         4.7404       <1H OCEAN  
272         555.0       190.0         3.7262        NEAR BAY  
8079        954.0       312.0         5.2606      NEAR OCEAN

In [12]:
# extract the labels corresponding to the test_data
test_label = y.loc[test_data.index]
test_label

18330    500001.0
3139      87900.0
11553    207500.0
272      166900.0
8079     198900.0
Name: median_house_value, dtype: float64

We pass the `test_data` through the pipeline. \
**Note** that this time we call the `.transform()` method because we called the `.fit()` method before.

In [13]:
test_data_prepared = full_pipeline.transform(test_data)
test_data_prepared

array([[-1.27528855,  0.84566614,  1.53876638, -0.26120196, -0.57637249,
        -0.63890851, -0.5800845 ,  2.53348067,  0.83711916, -0.0507843 ,
        -1.20110476,  0.        ,  0.        ,  0.        ,  1.        ,
         0.        ],
       [ 0.70417607, -0.3618595 , -1.55595157, -0.52194185, -0.58591916,
        -0.58525959, -0.58270947, -0.45286096, -0.00602006, -0.03105363,
        -0.31782733,  0.        ,  1.        ,  0.        ,  0.        ,
         0.        ],
       [ 0.79891115, -0.89073701,  0.03108328,  0.36834464,  0.23032085,
         0.73045819,  0.31503099,  0.4514386 ,  0.04942431,  0.04692891,
        -0.49015367,  1.        ,  0.        ,  0.        ,  0.        ,
         0.        ],
       [-1.30021884,  1.00011709,  1.85617335, -0.72290012, -0.82458583,
        -0.76643464, -0.81370702, -0.08116338,  0.08224509, -0.01519283,
        -0.56013289,  0.        ,  0.        ,  0.        ,  1.        ,
         0.        ],
       [ 0.6892179 , -0.85329435,  1

In [14]:
# predicting
predicted_data = LR_model.predict(test_data_prepared)
predicted_data

array([446399.443621  ,  98741.72145397, 234414.26188318, 234693.65804407,
       295036.94239139])

What you see these are the predicted values. Let's compare how they differ from actual values:

In [15]:
pd.DataFrame({
    'Prediction': predicted_data,
    'Real price': test_label
})

Prediction  Real price
18330  446399.443621    500001.0
3139    98741.721454     87900.0
11553  234414.261883    207500.0
272    234693.658044    166900.0
8079   295036.942391    198900.0

### 4\. Evaluating the Model

In [16]:
test_set

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
20046    -119.01     36.06                25.0       1505.0             NaN   
3024     -119.46     35.14                30.0       2943.0             NaN   
15663    -122.44     37.80                52.0       3830.0             NaN   
20484    -118.72     34.28                17.0       3051.0             NaN   
9814     -121.93     36.62                34.0       2351.0             NaN   
...          ...       ...                 ...          ...             ...   
15362    -117.22     33.36                16.0       3165.0           482.0   
16623    -120.83     35.36                28.0       4323.0           886.0   
18086    -122.05     37.31                25.0       4111.0           538.0   
2144     -119.76     36.77                36.0       2507.0           466.0   
3665     -118.37     34.22                17.0       1787.0           463.0   

       population  households  median_income  median_house_value  \
20046      1392.0       359.0         1.6812             47700.0   
3024       1565.0       584.0         2.5313             45800.0   
15663      1310.0       963.0         3.4801            500001.0   
20484      1705.0       495.0         5.7376            218600.0   
9814       1063.0       428.0         3.7250            278000.0   
...           ...         ...            ...                 ...   
15362      1351.0       452.0         4.6050            263300.0   
16623      1650.0       705.0         2.7266            266800.0   
18086      1585.0       568.0         9.2298            500001.0   
2144       1227.0       474.0         2.7850             72300.0   
3665       1671.0       448.0         3.5521            151500.0   

      ocean_proximity  
20046          INLAND  
3024           INLAND  
15663        NEAR BAY  
20484       <1H OCEAN  
9814       NEAR OCEAN  
...               ...  
15362       <1H OCEAN  
16623      NEAR OCEAN  
18086       <1H OCEAN  
2144           INLAND  
3665        <1H OCEAN  

[4128 rows x 10 columns]

In [17]:
# separating predictors
X_test = test_set.drop('median_house_value', axis=1)
X_test

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
20046    -119.01     36.06                25.0       1505.0             NaN   
3024     -119.46     35.14                30.0       2943.0             NaN   
15663    -122.44     37.80                52.0       3830.0             NaN   
20484    -118.72     34.28                17.0       3051.0             NaN   
9814     -121.93     36.62                34.0       2351.0             NaN   
...          ...       ...                 ...          ...             ...   
15362    -117.22     33.36                16.0       3165.0           482.0   
16623    -120.83     35.36                28.0       4323.0           886.0   
18086    -122.05     37.31                25.0       4111.0           538.0   
2144     -119.76     36.77                36.0       2507.0           466.0   
3665     -118.37     34.22                17.0       1787.0           463.0   

       population  households  median_income ocean_proximity  
20046      1392.0       359.0         1.6812          INLAND  
3024       1565.0       584.0         2.5313          INLAND  
15663      1310.0       963.0         3.4801        NEAR BAY  
20484      1705.0       495.0         5.7376       <1H OCEAN  
9814       1063.0       428.0         3.7250      NEAR OCEAN  
...           ...         ...            ...             ...  
15362      1351.0       452.0         4.6050       <1H OCEAN  
16623      1650.0       705.0         2.7266      NEAR OCEAN  
18086      1585.0       568.0         9.2298       <1H OCEAN  
2144       1227.0       474.0         2.7850          INLAND  
3665       1671.0       448.0         3.5521       <1H OCEAN  

[4128 rows x 9 columns]

In [18]:
# separating labels
y_test = test_set['median_house_value'].copy()
y_test

20046     47700.0
3024      45800.0
15663    500001.0
20484    218600.0
9814     278000.0
           ...   
15362    263300.0
16623    266800.0
18086    500001.0
2144      72300.0
3665     151500.0
Name: median_house_value, Length: 4128, dtype: float64

In [19]:
# pass test_set through pipeline
X_test_prepared = full_pipeline.transform(X_test)
X_test_prepared

array([[ 0.28534728,  0.1951    , -0.28632369, ...,  0.        ,
         0.        ,  0.        ],
       [ 0.06097472, -0.23549054,  0.11043502, ...,  0.        ,
         0.        ,  0.        ],
       [-1.42487026,  1.00947776,  1.85617335, ...,  0.        ,
         1.        ,  0.        ],
       ...,
       [-1.23041404,  0.78014149, -0.28632369, ...,  0.        ,
         0.        ,  0.        ],
       [-0.08860699,  0.52740357,  0.58654547, ...,  0.        ,
         0.        ,  0.        ],
       [ 0.60445493, -0.66608108, -0.92113763, ...,  0.        ,
         0.        ,  0.        ]])

In [20]:
# Prediction
y_predicted = LR_model.predict(X_test_prepared)

We use **Root Mean Square Error** (RMSE) to compare prediction and real data:

In [21]:
# Evaluation
from sklearn.metrics import mean_squared_error

lin_mse = mean_squared_error(y_test, y_predicted)
# calculate RMSE
lin_rmse = np.sqrt(lin_mse)
print(lin_rmse)

72701.32600762133



So, `RMSE = $72701` came out. Not bad, but not good either. That is, our model makes an average error of `$72,000` when evaluating houses.

There is no single, universal solution to improve model accuracy. Things we can try:
- Finding better parameters
- Choosing a better model (algorithm).
- Collecting more information, etc.

We will try another model now.

---

#### DecisionTree

**`sklearn.tree.DecisionTreeRegressor`** - A decision tree regressor.

In [22]:
from sklearn.tree import DecisionTreeRegressor

Tree_model = DecisionTreeRegressor()
Tree_model.fit(X_prepared, y)

DecisionTreeRegressor()

In [23]:
# Prediction
y_predicted = Tree_model.predict(X_test_prepared)

# Evaluation
lin_mse = mean_squared_error(y_test, y_predicted)
# calculate RMSE
lin_rmse = np.sqrt(lin_mse)
print(lin_rmse)

71819.67686182908


It is not much different from the previous result.

---

#### RandomForest

**`sklearn.ensemble.RandomForestRegressor`** - A random forest regressor.

In [24]:
from sklearn.ensemble import RandomForestRegressor

RF_model = RandomForestRegressor()
RF_model.fit(X_prepared, y)

RandomForestRegressor()

In [25]:
# Prediction
y_predicted = RF_model.predict(X_test_prepared)

# Evaluation
lin_mse = mean_squared_error(y_test, y_predicted)
# calculate RMSE
lin_rmse = np.sqrt(lin_mse)
print(lin_rmse)

50002.10883490696


Better than previous result.

---

### 5\. Cross-validation

With cross-validation, we can divide the dataset into several parts and train & test the model several times using different parts of the dataset.

<img width=600px src="https://mathworks.com/discovery/cross-validation/_jcr_content/mainParsys/image.adapt.full.medium.jpg/1670430048960.jpg" alt='cross-validation.jpg'>\
© https://mathworks.com/discovery/cross-validation

For cross validation, it is not necessary to divide the data into train and test, it is done by `sklearn` itself.

In [26]:
X = df.drop('median_house_value', axis=1)
y = df['median_house_value'].copy()

X_prepared = full_pipeline.transform(X)

We create a simple function to display the validation results:

In [27]:
def display_scores(scores):
    print("Scores:", scores)
    print("Mean:", scores.mean())
    print("Std.dev:", scores.std())

#### LogisticRegression validation

In [28]:
# import cross validation score
from sklearn.model_selection import cross_val_score

scores = cross_val_score(LR_model, X_prepared, y, scoring="neg_mean_squared_error", cv=10)
LR_rmse_scores = np.sqrt(-scores)

display_scores(LR_rmse_scores)

Scores: [84188.51219065 61197.24357613 86752.24346334 62289.14292385
 80540.40041898 68919.39949642 52503.82940087 90910.07884989
 77674.67507925 53941.60539478]
Mean: 71891.71307941682
Std.dev: 13249.525989444988


#### DecisionTree validation

In [29]:
scores = cross_val_score(Tree_model, X_prepared, y, scoring="neg_mean_squared_error", cv=10)
LR_rmse_scores = np.sqrt(-scores)

display_scores(LR_rmse_scores)

Scores: [118786.28429695  72388.5322726   83599.45043162  75877.76362881
  89809.48647574  79481.87833757  69675.63507202 101194.18729703
  94316.64812298  75276.87692927]
Mean: 86040.6742864587
Std.dev: 14515.221578152792


#### RandomForest validation

In [30]:
scores = cross_val_score(RF_model, X_prepared, y, scoring="neg_mean_squared_error", cv=10)
LR_rmse_scores = np.sqrt(-scores)

display_scores(LR_rmse_scores)

Scores: [97989.83458372 48177.39126102 65136.42834547 56481.14534201
 60812.57998636 60538.21012727 47046.08690237 78992.6679687
 73965.1145346  49426.07972913]
Mean: 63856.55387806427
Std.dev: 15196.52669706573


### 6\. Saving the Model

`joblib` is faster in saving/loading large NumPy arrays, whereas `pickle` is faster with large collections of Python objects. Therefore, if your model contains large NumPy arrays (as the majority of models does), `joblib` should be faster.

#### Saving with `pickle`

In [31]:
import pickle

filename = 'RF_model.pkl'  # we can give any name to file and extension
with open(filename, 'wb') as file:
    pickle.dump(RF_model, file)

Loading the model:

In [32]:
with open('RF_model.pkl', 'rb') as file:
    model = pickle.load(file)

Let's test the model:

In [33]:
scores = cross_val_score(model, X_prepared, y, scoring="neg_mean_squared_error", cv=5)
LR_rmse_scores = np.sqrt(-scores)
display_scores(LR_rmse_scores)

Scores: [77996.59325089 63953.08934642 61209.92586003 80865.18573225
 62347.28691674]
Mean: 69274.41622126565
Std.dev: 8387.609372045397


#### Saving with `joblib`

`pip install joblib`

In [34]:
import joblib

filename = 'RF_model.jbl'  # we can give any name to file and extension
joblib.dump(RF_model, filename)

['RF_model.jbl']

Loading the model:

In [35]:
model = joblib.load('RF_model.jbl')

Testing the model:

In [36]:
scores = cross_val_score(model, X_prepared, y, scoring="neg_mean_squared_error", cv=5)
LR_rmse_scores = np.sqrt(-scores)

display_scores(LR_rmse_scores)

Scores: [77723.3643669  64421.06298793 61530.25499824 82830.51797994
 62389.17915362]
Mean: 69778.87589732655
Std.dev: 8772.841979299752


**Saving `pipeline` with `joblib`:**

In [37]:
filename = 'pipeline.jbl'
joblib.dump(full_pipeline, filename)

['pipeline.jbl']

In [38]:
print("Done!")

Done!
